Add support for setting CPU limits for workloads

eggfoobar opened the pull request:
Added the ability to support cpuquota and cpuperiod for specific workloads, this allows more fine tune control for containers that are assigned to those workloads to be set to a specific CPU quota.

Added a wrapper struct for annotation object to be used for information such as cpu milli cores to be used in the calculation for cpu quota.

Design Info:
[Enhancement Proposal](https://github.com/openshift/enhancements/pull/1546)

<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

<!--
/kind api-change
/kind bug
/kind ci
/kind cleanup
/kind dependency-change
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake
/kind other
-->

#### What this PR does / why we need it:

Currently the workloads feature does not take into account `cpuquota` for cpu limits. CPU limits just get ignored for the time as it currently stands, this feature adds support for cpu limits by exposing cpu quota configuration.

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Extend `workloads` config option to support CPU limits for containers that use the workloads annotations. This will expose `cpuperiod` and `cpuquota` options for a desired default value outside the norm. This feature will primarily be utilized to ensure that containers with workload annotations are correctly limited, the average use cases will not need to alter the default values.
```

kwilczynski commented:
/ok-to-test

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -135,14 +159,35 @@ func (w Workloads) workloadGivenActivationAnnotation(sboxAnnotations map[string]
+		period = defaultQuotaPeriod
+	}
+
+	// we then convert your milliCPU to a value normalized over a period
```
```suggestion
	// We then convert the milliCPU to a value normalized over a period.
```

eggfoobar replied:
Updated, thanks!

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +58,22 @@ type WorkloadConfig struct {
+// extra information that is passed down from the pod.
+type ResourcesAnnotation struct {
+	Resources
+	// Specifies the CPU limit in millicores this will be used to calculate the cpu quota.
```
```suggestion
	// Specifies the CPU limit in millicores. This will be used to calculate the CPU quota.
```

kwilczynski commented on pkg/config/template.go:
```diff
@@ -1334,7 +1334,7 @@ const templateStringCrioRuntimeWorkloads = `# The workloads table defines ways t
 # Note, the behavior of this table is EXPERIMENTAL and may change at any time.
 # Each workload, has a name, activation_annotation, annotation_prefix and set of resources it supports mutating.
-# The currently supported resources are "cpu" (to configure the cpu shares) and "cpuset" to configure the cpuset.
+# The currently supported resources are "cpuperiod" "cpuquota", "cpushares" and "cpuset" to configure the cpuset.
```
This sentence now reads as if all of these above are required to configure "cpuset"—is that the case?

eggfoobar replied:
Good catch, that should not be required, updated wording there to remove that phrasing.

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -156,19 +201,37 @@ func resourcesFromAnnotation(prefix, ctrName string, allAnnotations map[string]s
+		resources.CPUPeriod = defaultResources.CPUPeriod
+	}
+
+	// If a CPU Limit in Milli is supplied via the annotation, calculate quota with given CPU Period
```
```suggestion
	// If a CPU Limit in Milli is supplied via the annotation, calculate quota with the given CPU period.
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -156,19 +201,37 @@ func resourcesFromAnnotation(prefix, ctrName string, allAnnotations map[string]s
+		return fmt.Errorf("cpuquota(%d) can not be less than cpushares(%d)", r.CPUQuota, r.CPUShares)
+	}
+	if r.CPUPeriod < minQuotaPeriod {
+		return fmt.Errorf("cpuperiod(%d) can not be less than 1000 usecs", r.CPUPeriod)
```
```suggestion
		return errors.New("cpuperiod cannot be less than 1000 microseconds")
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -156,19 +201,37 @@ func resourcesFromAnnotation(prefix, ctrName string, allAnnotations map[string]s
-	_, err := cpuset.Parse(r.CPUSet)
-	return err
+	if r.CPUQuota < int64(r.CPUShares) {
+		return fmt.Errorf("cpuquota(%d) can not be less than cpushares(%d)", r.CPUQuota, r.CPUShares)
```
```suggestion
		return fmt.Errorf("cpuquota %d cannot be less than cpushares %d", r.CPUQuota, r.CPUShares)
```

kwilczynski commented:
@eggfoobar, this needs a release note.

kwilczynski commented:
/retitle WIP: Add support for setting CPU limits for workloads

eggfoobar commented:
Thanks for the input @kwilczynski ! I updated the release notes block to include the intention with this change

kwilczynski commented:
/close

kwilczynski commented:
/reopen

kwilczynski commented:
/retest-required

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
 	CPUShares uint64 `json:"cpushares,omitempty"`
+	// Specifies the CPU quota this pod is limited to.
+	CPUQuota int64 `json:"cpuquota,omitempty"`
+	// Specifies the CPU period to use for these workloads
```
```suggestion
	// Specifies the CPU period to use for these workloads.
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
 type Resources struct {
 	// Specifies the number of CPU shares this pod has access to.
 	CPUShares uint64 `json:"cpushares,omitempty"`
+	// Specifies the CPU quota this pod is limited to.
+	CPUQuota int64 `json:"cpuquota,omitempty"`
+	// Specifies the CPU period to use for these workloads
+	CPUPeriod uint64 `json:"cpuperiod,omitempty"`
```
I wonder if mentioning units here would be useful for a reference.

eggfoobar replied:
That's a good point, we should so there's no confusion. I'll add that in there

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
 	CPUSet string `json:"cpuset,omitempty"`
 }
 
+// ResourceAnnotation describes extra information that is not part of the CRIO config but is used to contain
+// extra information that is passed down from the pod.
```
```suggestion
// ResourceAnnotation describes extra information that is not part of the CRI-O config but is used to contain
// extra information that is passed down from the pod.
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
 type Resources struct {
 	// Specifies the number of CPU shares this pod has access to.
 	CPUShares uint64 `json:"cpushares,omitempty"`
+	// Specifies the CPU quota this pod is limited to, the units are in micro seconds.
```
```suggestion
	// Specifies the CPU quota this Pod is limited to in microseconds.
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
 	CPUShares uint64 `json:"cpushares,omitempty"`
+	// Specifies the CPU quota this pod is limited to, the units are in micro seconds.
+	CPUQuota int64 `json:"cpuquota,omitempty"`
+	// Specifies the CPU period to use for these workloads, the units are in micro seconds.
```
```suggestion
	// Specifies the CPU period this Pod will use in microseconds.
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
 type Resources struct {
 	// Specifies the number of CPU shares this pod has access to.
```
```suggestion
	// Specifies the number of CPU shares this Pod has access to.
```

kwilczynski commented on pkg/config/workloads.go:
```diff
@@ -46,10 +59,22 @@ type WorkloadConfig struct {
+	CPUQuota int64 `json:"cpuquota,omitempty"`
+	// Specifies the CPU period to use for these workloads, the units are in micro seconds.
+	CPUPeriod uint64 `json:"cpuperiod,omitempty"`
 	// Specifies the cpuset this pod has access to.
```
```suggestion
	// Specifies the cpuset this Pod has access to.
```

kwilczynski commented on pkg/config/template.go:
```diff
@@ -1334,7 +1334,7 @@ const templateStringCrioRuntimeWorkloads = `# The workloads table defines ways t
 # Note, the behavior of this table is EXPERIMENTAL and may change at any time.
 # Each workload, has a name, activation_annotation, annotation_prefix and set of resources it supports mutating.
-# The currently supported resources are "cpu" (to configure the cpu shares) and "cpuset" to configure the cpuset.
+# The currently supported resources are "cpuperiod" "cpuquota", "cpushares" and "cpuset". ("cpuperiod" and "cpuquota" are in micro seconds)
```
```suggestion
# The currently supported resources are "cpuperiod" "cpuquota", "cpushares" and "cpuset". The values for "cpuperiod" and "cpuquota" are denoted in microseconds.
```

eggfoobar replied:
This is clearer, thanks, updated!

kwilczynski commented:
@eggfoobar, I appreciate you taking care of all the little wording updates. Much appreciated.

kwilczynski commented:
/retest-required

eggfoobar commented:
> @eggfoobar, I appreciate you taking care of all the little wording updates. Much appreciated.

Oh no worries at all, wording especially docs should be clear, so much appreciated as well, thanks for the input!

kwilczynski commented:
/retest-required

kwilczynski commented:
/retest-required

eggfoobar commented:
/hold

Holding for e2e validation with k8s.

eggfoobar commented:
/unhold

### Validated locally.

Installed a cluster with Workload Partitioning enabled. Validated with the below deployments.

> Note:  Non workload partitioning was also validated to make sure the `cpuquota` and `cpushares` were correctly calculated. Results correctly lined up as expected.

### Tested New Additions

<details>
  <summary>The Given Deployment With Requests and Limits</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: busy-box-deployment
  namespace: custom-namespace
spec:
...
  template:
    metadata:
      annotations:
        target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
...
    spec:
      containers:
        - name: limitbox
 ...
          resources:
            requests:
              cpu: 35m
              memory: 50Mi
            limits:
              cpu: 55m
              memory: 50Mi
```
</details>

Was correctly updated, by our webhook to include the new fields in the annotations:
```yaml
      resources.workload.openshift.io/limitbox: '{"cpushares":35,"cpulimit":55}'
      target.workload.openshift.io/management: '{"effect":"PreferredDuringScheduling"}'
```

Investigating the host container with `sudo crictl inspect 5230a1e7f6a7b | jq .info.runtimeSpec.linux.resources.cpu`, showed the correct quota being applied.
```json
{
  "shares": 35,
  "quota": 5500,
  "period": 100000
}
```

### Tested Original Behavior

<details>
  <summary>The Given Deployment With Just Requests</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: busy-box-deployment
  namespace: custom-namespace
spec:
...
  template:
    metadata:
      annotations:
        target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
...
    spec:
      containers:
        - name: limitbox
 ...
          resources:
            requests:
              cpu: 35m
              memory: 50Mi
```
</details>

Was correctly updated, by our webhook to include just the `cpushares`:
```yaml
      resources.workload.openshift.io/limitbox: '{"cpushares":35}'
      target.workload.openshift.io/management: '{"effect":"PreferredDuringScheduling"}'
```

Investigating the host container with `sudo crictl inspect 5230a1e7f6a7b | jq .info.runtimeSpec.linux.resources.cpu`, showed the correct quota being applied, which in this case is the default 0.
```json
{
  "shares": 35,
  "quota": 0,
  "period": 100000
}
```

kwilczynski commented:
/assign eggfoobar

kwilczynski commented:
/retest required

kwilczynski commented:
/retest-required

kwilczynski commented:
@eggfoobar, thank you for going above and beyond with the testing! Much appreciated.

We have some linter complaints. I wonder if it would make sense to rebase against `main` too? I wonder if it would pull some of the test updates, etc., that we made in the meantime.

kwilczynski commented:
/retest-required

eggfoobar commented:
Missed a few linter errors, pushed up fixes

eggfoobar commented on pkg/config/workloads_test.go:
```diff
@@ -0,0 +1,236 @@
+		// When
+		err := resources.ValidateDefaults()
+		// Then
+		Expect(err).To(BeNil())
```
Lol I can't believe I missed this one, that's what I get for trying to sneak in the change before logging off, thanks updating.

kwilczynski commented:
/retest-required

eggfoobar commented:
/assign @haircommander

haircommander commented on pkg/config/workloads.go:
```diff
@@ -135,14 +159,35 @@ func (w Workloads) workloadGivenActivationAnnotation(sboxAnnotations map[string]
 
+// milliCPUToQuota converts milliCPU to CFS quota and period values.
+// Input parameters and resulting value is number of microseconds.
+func milliCPUToQuota(milliCPU, period int64) (quota int64) {
```
nit: can functions be defined below they're called?

eggfoobar replied:
Yeah, absolutely, updating

eggfoobar commented:
@haircommander After our discussion, I moved the `cpulimit` field to the crio configuration proper, when you have a moment, feel free to give this another look. Thanks!

haircommander commented:
/approve

LGTM, thanks @eggfoobar

kwilczynski commented:
/retest-required

kwilczynski commented:
/approve
/lgtm

kwilczynski commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
Hey @kwilczynski the integration test revealed that the default validation check was too strict and didn't account for zero values, I went ahead and updated the validation check and added unit tests to catch it as well.

kwilczynski commented:
/retest

eggfoobar commented:
Seems the recent failures are not related to this PR

`fedora-integration` failed with
```
# time="2024-03-29 18:24:38.033037100Z" level=fatal msg="create metrics endpoint on 127.0.0.1:40845: creating listener: listen tcp 127.0.0.1:40845: bind: address already in use" file="crio/main.go:309"
```

and `hel-critest` failed with trying to ssh into the host to run the test
```
+ ssh -l deadbeef -o ConnectTimeout=5 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=90 -o LogLevel=ERROR -i /tmp/secret/vpc-sshkey 35.239.38.232 -- 'cat > ${HOME}/cri-o.tar.gz'
ssh: connect to host 35.239.38.232 port 22: Connection refused
```

eggfoobar commented:
/retest-required

eggfoobar commented:
@kwilczynski We should be good with the ci now, thanks for your patience!

kwilczynski commented:
/approve
/lgtm

kwilczynski commented:
/test ci-rhel-critest

eggfoobar commented:
/retest-required

Failed job errors don't seem related, retrying.

kwilczynski commented:
> /retest-required
> 
> Failed job errors don't seem related, retrying.

@eggfoobar, the CI jobs can be a hit and miss, sadly. Thank you for your patience!

eggfoobar commented:
/cherry-pick release-1.29

kwilczynski commented:
@eggfoobar, certainly.

kwilczynski commented:
/cherry-pick release-1.29

kwilczynski commented:
@eggfoobar, do we also need this in 1.28?

eggfoobar commented:
Thanks so much @kwilczynski ! This should only be needed for 1.29, the intent is for OCP 4.16+ to utilize this feature for now.
